src/wallet/witness_address.cpp

```diff
@@ -0,0 +1,550 @@
+{
+    return valid_;
+}
+
+// Serializer.
+// ----------------------------------------------------------------------------
+
+// Returns the bech32 encoded witness address.
+std::string witness_address::bech32() const
+{
+    base32 bech32;
+    bech32.prefix = prefix_;
+    bech32.payload = to_chunk(witness_version_);
+
+    const auto converted = (witness_hash_ == null_hash ?
+        convert_bits(bech32_expanded_bit_size, bech32_contracted_bit_size, true,
+            to_chunk(hash_), 0) : convert_bits(bech32_expanded_bit_size,
+                bech32_contracted_bit_size, true, to_chunk(witness_hash_), 0));
+
+    extend_data(bech32.payload, converted);
+    return encode_base32(bech32);
+}
+
+// Returns the bech32 or base58 encoded witness address, depending on
+// the encoding.
+std::string witness_address::encoded() const
+{
+    switch (version_)
+    {
+        case witness_address::mainnet_base58_p2wpkh:
+        case witness_address::testnet_base58_p2wpkh:
+        {
+            witness_p2wpkh witness;
+            to_witness(witness, version_, witness_version_, hash_);
+            return encode_base58(witness);
+        }
+
+        case witness_address::mainnet_base58_p2wsh:
+        case witness_address::testnet_base58_p2wsh:
+        {
+            witness_p2wsh witness;
+            to_witness(witness, version_, witness_version_, witness_hash_);
+            return encode_base58(witness);
+        }
+
+        case witness_address::mainnet_bech32_p2wpkh:
+        case witness_address::mainnet_bech32_p2wsh:
+        case witness_address::testnet_bech32_p2wpkh:
+        case witness_address::testnet_bech32_p2wsh:
+            return bech32();
+
+        default:
+            return {};
+    }
+}
+
+// Accessors.
+// ----------------------------------------------------------------------------
+
+uint8_t witness_address::witness_version() const
+{
+    return witness_version_;
+}
+
+const hash_digest& witness_address::witness_hash() const
+{
+    return witness_hash_;
+}
+
+chain::script witness_address::output_script() const
+{
+    switch (version_)
+    {
+        case witness_address::mainnet_bech32_p2wpkh:
+        case witness_address::testnet_bech32_p2wpkh:
+        case witness_address::mainnet_base58_p2wpkh:
+        case witness_address::testnet_base58_p2wpkh:
+            return chain::script::to_pay_witness_key_hash_pattern(hash_);
+
+        case witness_address::mainnet_bech32_p2wsh:
+        case witness_address::testnet_bech32_p2wsh:
+        case witness_address::mainnet_base58_p2wsh:
+        case witness_address::testnet_base58_p2wsh:
+            return chain::script::to_pay_witness_script_hash_pattern(witness_hash_);
+
+        default:
+            return {};
+    }
+}
+
+// Methods.
+// ----------------------------------------------------------------------------
+
+/// static
+void witness_address::to_witness(witness_p2wpkh& out, uint8_t version,
+    uint8_t witness_version, short_hash hash)
```

Comment:
Avoid parameter copies by always using `const short_hash& hash` idiom). When a copy is required, make it internally for the sake of consistency.

Reply:
See other instances.